### PR TITLE
Parallel type checking for impl files with backing sig files 

### PR DESCRIPTION
**Pull request by nojaf**
Trying to pick up where https://github.com/dotnet/fsharp/pull/11152 left.

### CODE REVIEW

File: src/Compiler/Driver/ParseAndCheckInputs.fs [hunk 1]
```diff
@@ -1338,11 +1351,44 @@ let CheckClosedInputSetFinish (declaredImpls: CheckedImplFile list, tcState) =
...
+
+            (inputs, results)
+            ||> Array.zip
+            |> Array.mapi (fun i (input, (_, _, implOpt, _)) ->
+                match implOpt with
+                | None -> None
+                | Some impl ->
+                    match impl with
+                    | CheckedImplFile.CheckedImplFile (qualifiedNameOfFile = qualifiedNameOfFile
+                                                       contents = ModuleOrNamespaceContents.TMDefs []) ->
```
Comment (dsyme) [inline src/Compiler/Driver/ParseAndCheckInputs.fs:1376]: Why the empty contents here?

Comment (dsyme) [reply src/Compiler/Driver/ParseAndCheckInputs.fs:1376]: I guess it matches the results produced by this line https://github.com/dotnet/fsharp/pull/13737/files#diff-efdad7fa53b23cfadc0c316627cd7f8833d0a531db500cc7c220cc4fa0465ab2R1227 but it feels like it would be better to return a more explicit marker, in case some input with a signature file also happens to have empty defs

**Comment by dsyme** [comment]
@nojaf Any idea what actually goes wrong here?

**Comment by nojaf** [comment]
@dsyme for some reason
```
C:\Users\nojaf\Projects\fsharp\src\Compiler\Checking\InfoReader.fs(932,5): error FS0971: Undefined value 'ExcludeHiddenOfMethInfos' [C:\Users\nojaf\Projects\fsharp\src\Compiler\FSharp.Compiler.Service.fsproj]
```
is no longer recognised.
At first glance, I'm not quite sure why. I'll try and reproduce this on a smaller scale.

**Comment by nojaf** [comment]
@dsyme I can reproduce it with the following:

signature:
```fsharp
module FSharpPlaygroundProject.InfoReader

type  TcGlobals = { A: int }

type  ImportMap = { A: int }

type  range = { A: int }

type  MethInfo = { A: int }

/// An InfoReader is an object to help us read and cache infos.
/// We create one of these for each file we typecheck.
type InfoReader =

    new: g: TcGlobals * amap: ImportMap -> InfoReader

    static member ExcludeHiddenOfMethInfos:
        g: TcGlobals -> amap: ImportMap -> m: range -> minfos: MethInfo list list -> MethInfo list

val  ExcludeHiddenOfMethInfos:
    g: TcGlobals -> amap: ImportMap -> m: range -> minfos: MethInfo list list -> MethInfo list
```

implementation:
```fsharp
module  FSharpPlaygroundProject.InfoReader

type TcGlobals = { A: int }
type ImportMap = { A: int }
type range = { A: int }
type MethInfo = { A: int }

/// An InfoReader is an object to help us read and cache infos.
/// We create one of these for each file we typecheck.
type InfoReader(g: TcGlobals, amap: ImportMap) =
    static member ExcludeHiddenOfMethInfos
        (g: TcGlobals)
        (amap: ImportMap)
        (m: range)
        (minfos: MethInfo list list)
        : MethInfo list =
        []

let ExcludeHiddenOfMethInfos (g: TcGlobals) (amap: ImportMap) (m: range) (minfos: MethInfo list list) : MethInfo list =
    InfoReader.ExcludeHiddenOfMethInfos g amap m minfos
```

The problem goes away when `ExcludeHiddenOfMethInfos` is renamed to `ExcludeHiddenOfMethInfos2`.
So I believe it to be a naming clash between a static member and a top-level binding.

**Comment by dsyme** [comment]
@nojaf I believe the problem is that all the signatures (e.g. of InfoReader.fsi) are present in the tcState at this line https://github.com/dotnet/fsharp/pull/13737/files#diff-efdad7fa53b23cfadc0c316627cd7f8833d0a531db500cc7c220cc4fa0465ab2R1384-R1387

Instead, the tcState and other inputs to use need to be exactly the ones that would have been used here https://github.com/dotnet/fsharp/pull/13737/files#diff-efdad7fa53b23cfadc0c316627cd7f8833d0a531db500cc7c220cc4fa0465ab2R1227

It is probably better to replace [these lines](https://github.com/dotnet/fsharp/pull/13737/files#diff-efdad7fa53b23cfadc0c316627cd7f8833d0a531db500cc7c220cc4fa0465ab2R1227
) with some more explicit return of the delayed computation to be processed later, including the relevant tcState etc.  Plus pthis entire section](https://github.com/dotnet/fsharp/pull/13737/files#diff-efdad7fa53b23cfadc0c316627cd7f8833d0a531db500cc7c220cc4fa0465ab2R1245-R1290) should likely be factored out as a separate function representing "work to be done to incorporate the results once the check of the impl file is completed".  I can pair program a draft with you if you like.

**Comment by dsyme** [comment]
@nojaf we should turn it on by default in the PR for testing purposes

**Comment by dsyme** [comment]
@nojaf Bootstrap now completes ✔️ 

I did a quick performance check on my machine after integrating https://github.com/dotnet/fsharp/pull/13740#issuecomment-1221376270

* Clean `build -noVisualStudio` before: 112sec (as reported in https://github.com/dotnet/fsharp/pull/13740)

* Clean `build -noVisualStudio` after: 92sec 

That ignores the time for the Proto phase. Note clean builds hit #13730 so are a PITA until that's fixed.

That's looking very promising given that 
* quite a lot of the files being compiled don't have signature files (FSharp.Build, most test files) 
* there's also significant fsyacc and fslex work in there

**Comment by nojaf** [comment]
Wonderful news! I'm having similar results when building the Fantomas solution (a couple of projects have signatures, some don't have any)

Before: 31sec
After: 18.40sec

**Comment by dsyme** [comment]
I pushed some more fixes regarding diagnostics 

There's quite a lot of cleanup in this branch now, we could factor it out, though I'm actually happy for the thing to go in as disabled or preview-enabled once green, I'll talk to @vzarytovskii and @KevinRansom about it.

Getting the diagnostics right is a bit tricky - the compiler uses this checkForErrors thing to prevent cascading errors for multiple phases of checking a file, and also from file to file.  So an error in an earlier file (e.g. signature file) can cause some errors to be suppressed in later files.   This makes it hard to get perfect match in diagnostics produced.  But what's there is very close.

**Comment by dsyme** [comment]
@nojaf I could see this causing a bunch of features....

* "add signatures for all files" action in IDEs
* IDE assists to maintain signatures under change to implementation
* IDE assists aligning comments and other metadata in signature and implementation
* Enable incremental compilation if the inferred signature of an implementation file hasn't changed (hard - requires intra-compilation state)

**Comment by nojaf** [comment]
Yes, generating all signatures correctly is already on my radar:
https://github.com/dotnet/fsharp/issues?q=is%3Aissue+label%3AArea-Compiler-SigFileGen+author%3Anojaf

**Comment by dsyme** [comment]
@nojaf The failures now look mostly related to the fact that we delay formatting of exceptions until "after the whole file has been checked", which means more type inference parameters have been solved.

I need to have a think about the best way to approach this.  We use the delay-and-replay-diagnostics in other places but this is the first time we're doing it for anything involving type inference.  I guess we need to format the diagnostics early.

**Comment by dsyme** [comment]
@nojaf I guess people would also love some kind of  "store local signature files silently, automatically keep them up to date each time I compile, use them for better compilation performance if they're there, if something changes then update them and re-do compilation" feature.

**Comment by dsyme** [comment]
@nojaf I guess people would also love some kind of  "store local signature files silently, automatically keep them up to date each time I compile, use them for better compilation performance if they're there, if something changes then update them and re-do compilation" feature.

**Comment by nojaf** [comment]
>  people would also love some kind of "store local signature files silently, automatically keep them up to date each time I compile, use them for better compilation performance if they're there, if something changes then update them and re-do compilation" feature.

Yeah, I thought about that, but I think you typically will want to generate the signatures once.
Throw out what you don't want to expose and keep impl/sig in sync afterwards.

As soon as the implementation no longer matches the signature you want to `alt-enter` to sync the signature file. That is at least the approach I'm thinking of.

**Comment by nojaf** [comment]
> Determinism failed for the following binaries:

Would any of this have an impact on the stability of the mvid in references assemblies?

**Comment by kerams** [comment]
> Yeah, I thought about that, but I think you typically will want to generate the signatures once.
Throw out what you don't want to expose and keep impl/sig in sync afterwards.

I imagine that would be geared towards users that don't need or care about signature files (like me), but would like to reap the performance benefits of having them. I mean, why not have the compiler always generate hidden signature files (unless explicit signatures exist) in the bin folder? Sort of like ref assemblies, but on the file level.

**Comment by auduchinok** [comment]
> I mean, why not have the compiler always generate hidden signature files (unless explicit signatures exist) in the bin folder?

Because it'll break how inferred type changes flow between files.

**Comment by Lenne231** [comment]
> > I mean, why not have the compiler always generate hidden signature files (unless explicit signatures exist) in the bin folder?
> 
> Because it'll break how inferred type changes flow between files.

Would it be possible to enforce explicit signatures of public APIs in fs files with a compiler flag? A fs file could be treated similar to a signature file then, right?

**Comment by auduchinok** [comment]
> Would it be possible to enforce explicit signatures of public APIs in fs files with a compiler flag? A fs file could be treated similar to a signature file then, right?

This is what I've also had in mind. It could work and would probably be much more pleasant than requiring signature files. @dsyme have you considered that before?

**Comment by kerams** [comment]
> Because it'll break how inferred type changes flow between files.

Couldn't we also invalidate the signature files of all dependent files?

**Comment by dsyme** [comment]
@auduchinok @kerams @Lenne231 let's take the discussion of compilation and tooling related to signature files to a discussion thread or similar.

@nojaf 

> Would any of this have an impact on the stability of the mvid in references assemblies?

I'm not sure.  Most likely it's non-determinism in generation of compiler-generated names of some kind.

**Comment by dsyme** [comment]
@nojaf I think I've fixed the remaining errors.   The core of the fix is to do eager formatting on diagnostics as they are produced in the parallel type checking. 

There's another set of code simplifications in the latest commits:

* SplitRelatedDiagnostics was old code that never actually returned any "related" diagnostics, so the second list was always empty.  The routine actually basically doesn nothing and should likely be removed.

* The "DiagnosticLoggerProvider" stuff was over-complicated and could be simplified out

**Comment by dsyme** [comment]
@nojaf I expect the detminism bugs to be just about the only ones left.  Could you run through the test scipt as used in CI and get a sample of the changes, hopefully revealed by using ILDASM?  I think it just builds twice and compares results.

https://github.com/dotnet/fsharp/blob/main/azure-pipelines.yml#L227

**Comment by vzarytovskii** [comment]
> @nojaf I expect the detminism bugs to be just about the only ones left.  Could you run through the test scipt as used in CI and get a sample of the changes, hopefully revealed by using ILDASM?  I think it just builds twice and compares results.
> 
> https://github.com/dotnet/fsharp/blob/main/azure-pipelines.yml#L227

It does compile twice with changing path, and then comparing assemblies on by one.

**Comment by nojaf** [comment]
@dsyme `StripRelatedDiagnostics` is still being used in `CollectFormattedDiagnostics`.

```
C:\Users\nojaf\Projects\fsharp\src\Compiler\Driver\CompilerDiagnostics.fs(2036,24): error FS0039: The value or constructor 'StripRelatedDiagnostics' is not defined. Maybe you want one of the following:↔   stripDebugPoints [C:\Users\nojaf\Projects\fsha
rp\src\Compiler\FSharp.Compiler.Service.fsproj]
C:\Users\nojaf\Projects\fsharp\src\Compiler\Driver\CompilerDiagnostics.fs(2044,44): error FS0072: Lookup on object of indeterminate type based on information prior to this program point. A type annotation may be needed prior to this program point to c
onstrain the type of the object. This may allow the lookup to be resolved. [C:\Users\nojaf\Projects\fsharp\src\Compiler\FSharp.Compiler.Service.fsproj]
```

What would the alternative be there?

**Comment by vzarytovskii** [comment]
`StripRelatedDiagnostics` is missing here, probably was added instead of deleted `SplitRelatedDiagnostics`.
Should probably be as easy as returning a phased diagnostics from the splitted exceptions.

**Comment by dsyme** [comment]
@nojaf I had to add some error recovery for combining fragments of namespaces because of a slight difference in paths between parallel and non-parallel when this fails. This might cause a few further failues

Regarding determinism, it turns out the artifacts published for the determinism build include the two relevant DLLs that have differences, e.g. deep in the zip available via here:

image

then here:

image

I'm comparing the ILdasm now

**Comment by dsyme** [comment]
Sample places the diffs differ:

image
image

There are about 100 other places that look identical.  Once we track down this non-determinism in name generation (Resulting no doubt from a global name generator used during type checking) I suspect this will pass

**Comment by dsyme** [comment]
And here it is:

```
        ident(g.CompilerGlobalState.Value.NiceNameGenerator.FreshCompilerGeneratedName("init", m), m)
```

**Comment by dsyme** [comment]
@nojaf I expect a swathe of baselines will need to be updated after this.

**Comment by nojaf** [comment]
> @nojaf I expect a swathe of baselines will need to be updated after this.

@dsyme I guess not. This looks promising. Could you summarize what still needs to be done here in order to merge this one?

**Comment by dsyme** [comment]
@nojaf Yay!

**Comment by dsyme** [comment]
We should turn this off by default and get it in.  Will discuss with @vzarytovskii

**Comment by vzarytovskii** [comment]
Yeah, this close to release it should be off by default.

**Comment by nojaf** [comment]
@dsyme @vzarytovskii how do you envision this? A new compiler flag?

**Comment by vzarytovskii** [comment]
> @dsyme @vzarytovskii how do you envision this? A new compiler flag?

I guess new TcConfig option which is enabled either by FCS API (by FSAC or VS via settings) or by the compiler flag for the fcs.dll.

**Comment by nojaf** [comment]
Well, as you want this improvement both in the editor and during a regular build, I assume a compiler flag would be the best fit. Any preference on the flag name?

**Comment by vzarytovskii** [comment]
> Well, as you want this improvement both in the editor and during a regular build, I assume a compiler flag would be the best fit. 

I mean it should be all three, I guess - TcConfig option backed by compiler flag (to use in plain `dotnet build`), and settable by FCS API (via `FSharpProjectOptions`?) so other consumers (Rider, FSAC, VS) can set it as part of language service setup.

> Any preference on the flag name?
No, no preferences really.

**Comment by auduchinok** [comment]
> and settable by FCS API (via FSharpProjectOptions?)

There's probably no reason to have this setting set to different values between different projects. Maybe add a `FSharpChecker` property or a parameter in its `Create`?

**Comment by vzarytovskii** [comment]
> > and settable by FCS API (via FSharpProjectOptions?)
> 
> There's probably no reason to have this setting set to different values between different projects. Maybe add a `FSharpChecker` property or a parameter in its `Create`?

Yeah, makes sense.

**Comment by nojaf** [comment]
@vzarytovskii, @auduchinok thanks, I made the first attempt. There is probably a bit more to it.
Let me know where I can improve.

**Comment by nojaf** [comment]
Yeah, we've seen some improvements around 17% - 19% faster.
https://github.com/dotnet/fsharp/pull/13737#issuecomment-1223386853
https://github.com/dotnet/fsharp/pull/13737#issuecomment-1223637818

**Comment by auduchinok** [comment]
Would `parallelTypeChecking` name still work well if we enable the parallel type checking that is discussed in https://github.com/dotnet/fsharp/discussions/11634? There's a chance we'd want to have separate settings for these two optimizations, at least while they are considered experimental.

Maybe a less generic name that would mention backing signature files would work better here? Though, it may be not easy to find a name that is both descriptive and concise 🙂

**Comment by baronfel** [comment]
I think a longer, more descriptive name would be fine here - FSharpChecker is not created very often, and the choices here have large impacts. Some of the other parameters already have quite long and descriptive names. Maybe something like `enable ParallelCheckingWithSignatureFiles` (and similar for the CompilerOption)?

**Comment by dsyme** [comment]
I actually think this should go in off-by-default but with an environment variable or completely hidden `--test:ParallelTypeChecking` flag at first, plus perhaps a CI run to enable.

We put it in as soon as we branch away from dev17.4

We then turn it on after a few weeks if we've been regularly using it.  

@vzarytovskii What do you think?

**Comment by vzarytovskii** [comment]
> I actually think this should go in off-by-default but with an environment variable or completely hidden `--test:ParallelTypeChecking` flag at first, plus perhaps a CI run to enable.
> 

Yes, we can enable it in CI by default. Maybe even for compiler solution

> We put it in as soon as we branch away from dev17.4
> 
You mean 17.5? We can put it earlier, but we want to make sure that by default we use the old way of typechecking.

> We then turn it on after a few weeks if we've been regularly using it.
> 
> @vzarytovskii What do you think?

Yep, we also should have a VS flag which will turn it on and off for vs tooling.

**Comment by dsyme** [comment]
This is ready I think

File: src/Compiler/Driver/ParseAndCheckInputs.fs [hunk 2]
```diff
@@ -1338,11 +1428,138 @@ let CheckClosedInputSetFinish (declaredImpls: CheckedImplFile list, tcState) =
...
+                tcsCreatesGeneratedProvidedTypes = tcState.tcsCreatesGeneratedProvidedTypes || createsGeneratedProvidedTypes
+            }
+
+        let finishedResults =
+            unfinishedResults
+            |> List.map (function
+                | Choice1Of2 result -> result
+                | Choice2Of2 (result, _) -> result)
+
+        finishedResults, tcState)
```
Comment (0101) [inline src/Compiler/Driver/ParseAndCheckInputs.fs:1554]: This could be simplified a bit by not re-wrapping the results in `Chioce`. 

```suggestion
        let results, createsGeneratedProvidedTypesFlags =

            List.zip partialResults inputsWithLoggers
            |> List.toArray
            |> ArrayParallel.map (fun (partialResult, (_, logger)) ->
                use _ = UseDiagnosticsLogger logger
                use _ = UseBuildPhase BuildPhase.TypeCheck

                RequireCompilationThread ctok

                match partialResult with
                | Choice1Of2 result -> result, false
                | Choice2Of2 (amap, conditionalDefines, rootSig, priorErrors, file, tcStateForImplFile, ccuSigForFile) ->

                    // In the first linear part of parallel checking, we use a 'checkForErrors' that checks either for errors
                    // somewhere in the files processed prior to this one, including from the first phase, or in the processing
                    // of this particular file.
                    let checkForErrors2 () = priorErrors || (logger.ErrorCount > 0)

                    let topAttrs, implFile, tcEnvAtEnd, createsGeneratedProvidedTypes =
                        CheckOneImplFile(
                            tcGlobals,
                            amap,
                            tcStateForImplFile.tcsCcu,
                            tcStateForImplFile.tcsImplicitOpenDeclarations,
                            checkForErrors2,
                            conditionalDefines,
                            TcResultsSink.NoSink,
                            tcConfig.internalTestSpanStackReferring,
                            tcStateForImplFile.tcsTcImplEnv,
                            Some rootSig,
                            file
                        )
                        |> Cancellable.runWithoutCancellation

                    let result = (tcEnvAtEnd, topAttrs, Some implFile, ccuSigForFile)
                    result, createsGeneratedProvidedTypes)
            |> Array.toList
            |> List.unzip

        let tcState =
            { tcState with
                tcsCreatesGeneratedProvidedTypes = tcState.tcsCreatesGeneratedProvidedTypes || (createsGeneratedProvidedTypesFlags |> List.exists id)
            }

        results, tcState)
```

**Review by vzarytovskii — CHANGES_REQUESTED** [review]
Probably a suggested CI leg with `--test:ParallelCheckingWithSignatureFiles` should be added (to the plain builds maybe?).

File: src/Compiler/Driver/ParseAndCheckInputs.fs [hunk 3]
```diff
@@ -1216,91 +1287,110 @@ let CheckOneInput
...
-                        |> cancellable.Return
-                    else
+                match rootSigOpt with
+                | Some rootSig when skipImplIfSigExists ->
+                    // Delay the typecheck the implementation file until the second phase of parallel processing.
+                    // Adjust the TcState as if it has been checked, which makes the signature for the file available later
+                    // in the compilation order.
+                    let tcStateForImplFile = tcState
+                    let qualNameOfFile = file.QualifiedName
+                    let hadSig = true
```
Comment (TIHan) [inline src/Compiler/Driver/ParseAndCheckInputs.fs:1297]: `hadSig` would already be true at this point. See line just above:
```fsharp
                let hadSig = rootSigOpt.IsSome
```

File: src/Compiler/Driver/fsc.fsi [hunk 4]
```diff
@@ -13,18 +13,18 @@ open FSharp.Compiler.CodeAnalysis
 open FSharp.Compiler.Syntax
 open FSharp.Compiler.TcGlobals
 
-[<AbstractClass>]
+/// DiagnosticLoggers can be sensitive to the TcConfig flags. During the checking
+/// of the flags themselves we have to create temporary loggers, until the full configuration is
+/// available.
 type DiagnosticsLoggerProvider =
```
Comment (TIHan) [inline src/Compiler/Driver/fsc.fsi:19]: nit: Since this is now an interface, we should prefix it with `I` -> `IDiagnosticsLoggerProvider` .

**Review by vzarytovskii — APPROVED** [review]

**Comment by nojaf** [comment]
Hello @vzarytovskii and @dsyme, any thoughts on how we can move forward with this?
I understand you probably don't want this for `17.4` but could we merge it into `main` maybe?

**Comment by nojaf** [comment]
Many thanks @vzarytovskii for merging this in!

**Comment by TIHan** [comment]
Congrats @nojaf for your work on this! Congrats to the F# team on getting this in!

**Comment by nojaf** [comment]
Thank you @TIHan, although I can hardly take any credit here. The original idea came from yourself, @dsyme has been a tremendous help during the implementation and the F# team supported us to get in it.
I'm very excited about this either way 🎉.
